src/context.ts

```diff
@@ -0,0 +1,236 @@
+ * from available HTTP header context.
+ *
+ * Specifically:
+ * - We extract trace context if `traceparent` is available via W3 protocol
+ * - We extract trace context if `X-Cloud-Trace-Context` is available from the
+ * user or Google Cloud Trace.
+ * - In the middleware scenario, we auto-inject a Google trace context if one
+ * is not available.
+ */
+
+import * as http from 'http';
+import * as uuid from 'uuid';
+import * as crypto from 'crypto';
+
+/** Header that carries span context across Google infrastructure. */
+export const X_CLOUD_HEADER = 'x-cloud-trace-context';
+const SPAN_ID_RANDOM_BYTES = 8;
+const spanIdBuffer = Buffer.alloc(SPAN_ID_RANDOM_BYTES);
+const randomFillSync = crypto.randomFillSync;
+const randomBytes = crypto.randomBytes;
+const spanRandomBuffer = randomFillSync
+  ? () => randomFillSync(spanIdBuffer)
+  : () => randomBytes(SPAN_ID_RANDOM_BYTES);
+
+/** Header that carries span context across W3C compliant infrastructure. */
+export const TRACE_PARENT_HEADER = 'traceparent';
+
+/**
+ * An transport and environment neutral API for getting request headers.
+ */
+export interface HeaderWrapper {
+  getHeader(name: string): string | string[] | undefined;
+  setHeader(name: string, value: string): void;
+}
+
+/**
+ * makeHeaderWrapper returns a wrapper with set and get header functionality,
+ * returning null if the incoming request object doesn't contain the 'header'
+ * propery.
+ * @param req
+ */
+export function makeHeaderWrapper(
+  req: http.IncomingMessage
+): HeaderWrapper | null {
+  if (!req.headers) return null;
+  const wrapper = {
+    setHeader(name: string, value: string) {
+      req.headers[name] = value;
+    },
+    getHeader(name: string) {
+      return req.headers[name];
+    },
+  };
+  return wrapper;
+}
+
+/**
+ * CloudTraceContext: Cloud Logging compliant trace context.
+ */
+export interface CloudTraceContext {
+  trace: string;
+  spanId?: string;
+  traceSampled?: boolean;
+}
+
+/**
+ * getOrInjectContext returns a CloudTraceContext with as many available trace
+ * and span properties as possible. It examines HTTP headers for trace context.
+ * Optionally, it can inject a Google compliant trace context when no context is
+ * available from headers.
+ *
+ * @param req
+ * @param projectId
+ * @param inject
+ */
+export function getOrInjectContext(
+  req: http.IncomingMessage,
+  projectId: string,
+  inject?: boolean
+): CloudTraceContext {
+  const defaultContext = toCloudTraceContext({}, projectId);
+  const wrapper = makeHeaderWrapper(req);
+  if (wrapper) {
+    // Detect 'traceparent' header.
+    const traceContext = getContextFromTraceParent(wrapper, projectId);
+    if (traceContext) return traceContext;
+    // Detect 'X-Cloud-Trace-Context' header.
+    const cloudContext = getContextFromXCloudTrace(wrapper, projectId);
+    if (cloudContext) return cloudContext;
+    // Optional: Generate and inject a context for the user;
+    if (inject) {
```

Comment:
It looks like inject is set explicitly by a bool flag. Should it then be the prioritized option, since users explicitly asked for it?

Reply:
this function is invoked internally only and not exported to users. but good pt - ill make it clearer in comments that it is only meant to kick in as a last option